openhands/sdk/agent/base.py

```diff
@@ -26,81 +26,158 @@
-            tool_class = getattr(openhands.tools, tool_spec.name, None)
-            if tool_class is None:
-                raise ValueError(
-                    f"Unknown tool name: {tool_spec.name}. Not found in openhands.tools"
-                )
-            tool_or_tools = tool_class.create(**tool_spec.params)
-            if isinstance(tool_or_tools, list):
-                tools.extend(tool_or_tools)
-            else:
-                tools.append(tool_or_tools)
+    def initialize(self):
+        """Create an AgentBase instance from an AgentSpec."""
+        if self._tools:
+            logger.warning("Agent already initialized; skipping re-initialization.")
+            return
 
-        # Check tool types
-        for tool in tools:
-            if not isinstance(tool, ToolBase):
-                raise ValueError(
-                    f"Tool {tool} is not an instance of 'Tool'. Got type: {type(tool)}"
-                )
+        tools: list[Tool] = []
+        for tool_spec in self.tools:
+            tools.extend(resolve_tool(tool_spec))
 
         # Add MCP tools if configured
-        if spec.mcp_config:
-            mcp_tools = create_mcp_tools(spec.mcp_config, timeout=30)
+        if self.mcp_config:
+            mcp_tools = create_mcp_tools(self.mcp_config, timeout=30)
             tools.extend(mcp_tools)
 
         logger.info(
             f"Loaded {len(tools)} tools from spec: {[tool.name for tool in tools]}"
         )
-        if spec.filter_tools_regex:
-            pattern = re.compile(spec.filter_tools_regex)
+        if self.filter_tools_regex:
+            pattern = re.compile(self.filter_tools_regex)
             tools = [tool for tool in tools if pattern.match(tool.name)]
             logger.info(
                 f"Filtered to {len(tools)} tools after applying regex filter: "
                 f"{[tool.name for tool in tools]}",
             )
 
-        return cls(
-            llm=spec.llm,
-            agent_context=spec.agent_context,
-            tools=tools,
-            system_prompt_filename=spec.system_prompt_filename,
-            system_prompt_kwargs=spec.system_prompt_kwargs,
-            condenser=spec.condenser,
-        )
+        # Always include built-in tools; not subject to filtering
+        tools.extend(BUILT_IN_TOOLS)
+
+        # Check tool types
+        for tool in tools:
+            if not isinstance(tool, Tool):
+                raise ValueError(
+                    f"Tool {tool} is not an instance of 'Tool'. Got type: {type(tool)}"
+                )
+
+        # Check name duplicates
+        tool_names = [tool.name for tool in tools]
+        if len(tool_names) != len(set(tool_names)):
+            duplicates = set(name for name in tool_names if tool_names.count(name) > 1)
+            raise ValueError(f"Duplicate tool names found: {duplicates}")
+
+        # Store tools in a dict for easy access
+        self._tools = {tool.name: tool for tool in tools}
```

Comment:
This seems a bit brittle to me or at least worth investigating. The registry key is different from the actual tool name correct? 


So `filter_tools_regex` matches runtime Tool.name (like `executebash`), not the registry key `BashTool`?

Reply:
> The registry key is different from the actual tool name correct?

One registry key corresponds to one tool class, which could give you one or multiple tools (e.g., one MCP gives me a bunch of tools), so it is not possible to enforce the mapping there.

> So filter_tools_regex matches runtime Tool.name (like executebash), not the registry key BashTool?

Yes -- since registry key is more for "class name" that we used internally; for tools like MCP, we don't really get their name until execution time

Reply:
Ah gotcha! Let's document that `ToolSpec.name` is the registry key then not the actual tool name.